src/wallet/wallet-generation.ts

```diff
@@ -0,0 +1,198 @@
+        )
+      )
+    }
+    // And its address
+    address = wallet
+  }
+
+  // Based on the connection URL, retrieve the correct faucet URL
+  const faucetNetwork = getFaucetUrl(this)
+
+  return new Promise((resolve, reject) => {
+    axios
+      .post<FaucetWallet>(faucetNetwork, body)
+      .then(async (response) => {
+        address = response.data.account.classicAddress
+        try {
+          // Check at regular interval if the account is enabled on the XRPL and funded
+          const isAcctFunded = await hasAccountBalanceIncreased(
+            this,
+            address,
+            startingBalance
+          )
+
+          if (isAcctFunded) {
+            resolve(response.data)
+          } else {
+            reject(
+              new errors.XRPLFaucetError(
+                `Unable to fund address with faucet after waiting ${INTERVAL_SECONDS} seconds`
+              )
+            )
+          }
+        } catch (err) {
+          reject(new errors.XRPLFaucetError(err))
+        }
+      })
+      .catch((error) => {
+        reject(new errors.XRPLFaucetError(error.data))
+      })
+  })
+}
+
+/**
+ * Retrieves an XRPL address balance by currency
+ *
+ * @param api - RippleAPI
+ * @param address - Wallet address.
+ * @param currency - Currency to search, default XRP.
+ * @returns
+ */
+async function getWalletBalanceByCurrency(
+  api: RippleAPI,
+  address: string,
+  currency: string = 'XRP'
+): Promise<string> {
+  let curr = currency
+
+  // If issued currency longer than 3 characters
+  if (currency.length > 3) {
+    curr = Buffer.from(currency, 'ascii')
+      .toString('hex')
+      .toUpperCase()
+      .padEnd(40, '0')
+  }
+
+  try {
+    // Get all the account balances
+    const balances = await api.getBalances(address)
+
+    // Retrieve only the 'curr' (typically XRP) balance
+    const xrpCurrency = balances.filter((balance) => {
+      return balance.currency === curr
+    })
+
+    return xrpCurrency[0].value
+  } catch (err) {
+    reject(`Account ${address} does not exist`)
+  }
+}
+
+/**
+ * Check at regular interval if the account is enabled on the XRPL and funded
+ *
+ * @param api - RippleAPI
+ * @param address - the account address to check
+ * @param originalBalance - the initial balance before the funding
+ * @returns A Promise boolean
+ */
+async function hasAccountBalanceIncreased(
+  api: RippleAPI,
+  address: string,
+  originalBalance: number
+): Promise<boolean> {
+  return new Promise((resolve, reject) => {
+    let attempts = MAX_ATTEMPS
+    const interval = setInterval(async () => {
+      if (attempts < 0) {
+        clearInterval(interval)
+        reject(false)
+      } else {
+        attempts--
+      }
+
+      try {
+        const newBalance = +(await getWalletBalanceByCurrency(api, address))
+        if (newBalance > originalBalance) {
+          clearInterval(interval)
+          resolve(true)
+        }
+      } catch (err) {
+        reject(
```

Comment:
Should we also `clearInterval` here before we reject this promise.

Reply:
Or do we want it to keep retrying, even if it fails to retrieve balance one time.

Reply:
Thanks, that's a good input.
I did several tests and if the address balance isn't retrieved, the function `getAddressXrpBalance` handles the error. The `catch` block you mentioned here will not be triggered. 
It will be triggered if what happens after `const newBalance = +(await getWalletBalanceByCurrency(api, address))` creates an error.
So I think it's good to add `clearInterval` here.
Let me know if that doesn't make sense.